tokio-stream/src/wrappers/broadcast.rs

```diff
@@ -0,0 +1,77 @@
+use std::pin::Pin;
+use tokio::sync::broadcast::error::RecvError;
+use tokio::sync::broadcast::Receiver;
+
+use futures_core::Stream;
+use tokio_util::sync::ReusableBoxFuture;
+
+use std::fmt;
+use std::task::{Context, Poll};
+
+/// A wrapper around [`tokio::sync::broadcast::Receiver`] that implements [`Stream`].
+///
+/// [`tokio::sync::broadcast::Receiver`]: struct@tokio::sync::broadcast::Receiver
+/// [`Stream`]: trait@crate::Stream
+pub struct BroadcastStream<T> {
+    inner: ReusableBoxFuture<Result<(T, Receiver<T>), WrappedRecvError<T>>>,
+}
+
+/// An error returned from the inner stream of a [`BroadcastStream`].
+#[derive(Debug, PartialEq)]
+pub enum BroadcastStreamRecvError {
+    /// The receiver lagged too far behind. Attempting to receive again will
+    /// return the oldest message still retained by the channel.
+    ///
+    /// Includes the number of skipped messages.
+    Lagged(u64),
+}
+
+#[derive(Debug)]
+enum WrappedRecvError<T> {
+    Lagged(u64, Receiver<T>),
+    Closed,
+}
+
+async fn make_future<T: Clone>(
+    mut rx: Receiver<T>,
+) -> Result<(T, Receiver<T>), WrappedRecvError<T>> {
+    match rx.recv().await {
+        Ok(item) => Ok((item, rx)),
+        Err(RecvError::Lagged(n)) => Err(WrappedRecvError::Lagged(n, rx)),
+        Err(RecvError::Closed) => Err(WrappedRecvError::Closed),
+    }
+}
```

Comment:
We should be able to simplify to the following.
```suggestion
async fn make_future<T: Clone>(
    mut rx: Receiver<T>,
) -> (Result<T, RecvError<T>>, Receiver<T>) {
    let result = rx.recv().await;
    (result, rx)
}
```

Reply:
That looks much better 😄

Reply:
That looks much better 😄